tsdb/engine/tsm1/compact.go

```diff
@@ -350,10 +354,21 @@ func (c *DefaultPlanner) PlanOptimize() ([]CompactionGroup, int64) {
 	// split across several files in sequence.
 	generations := c.findGenerations(true)
 
-	// If there is only one generation and no tombstones, then there's nothing to
-	// do.
-	if len(generations) <= 1 && !generations.hasTombstones() {
-		return nil, 0
+	// Safety check for potential 0 generations.
+	if len(generations) < 1 {
+		return nil, 0, 0
+	}
+
+	// Return if there's only a single file and single generation.
+	if len(generations) == 1 && len(generations[0].files) == 1 {
+		return nil, 0, 0
+	}
+
+	// If there is a single generation, no tombstones, and the entire group size (all TSM files in the generation)
+	// is over or equal to 2 GB (Max TSM file size) there is nothing to do and we will return. Please see: https://github.com/influxdata/plutonium/issues/4201
```

Comment:
I am working on transferring my plutonium ticket https://github.com/influxdata/plutonium/issues/4201 to an OSS repo ticket so I can link to it in the code comments.

Reply:
I'd prefer to see the explanation of the logic in the code itself. If there is a reason it can't be in the code, link it to a specific comment in the issue. There is a lot to read in that issue!

Reply:
Sounds good can do